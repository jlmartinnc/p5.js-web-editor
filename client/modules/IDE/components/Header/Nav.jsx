import React, { useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { sortBy } from 'lodash';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import MenubarSubmenu from '../../../../components/Menubar/MenubarSubmenu';
import MenubarItem from '../../../../components/Menubar/MenubarItem';
import { availableLanguages, languageKeyToLabel } from '../../../../i18n';
import getConfig from '../../../../utils/getConfig';
import { showToast } from '../../actions/toast';
import { setLanguage } from '../../actions/preferences';
import Menubar from '../../../../components/Menubar/Menubar';
import CaretLeftIcon from '../../../../images/left-arrow.svg';
import LogoIcon from '../../../../images/p5js-logo-small.svg';
import { selectRootFile } from '../../selectors/files';
import { selectSketchPath } from '../../selectors/project';
import { metaKey, metaKeyName } from '../../../../utils/metaKey';
import { useSketchActions } from '../../hooks';
import { getAuthenticated, getIsUserOwner } from '../../selectors/users';
import { cloneProject } from '../../actions/project';
import {
  newFile,
  newFolder,
  showKeyboardShortcutModal,
  startSketch,
  stopSketch
} from '../../actions/ide';
import { logoutUser } from '../../../User/actions';
import { CmControllerContext } from '../../pages/IDEView';
import MobileNav from './MobileNav';
import useIsMobile from '../../hooks/useIsMobile';

const Nav = ({ layout }) => {
  const isMobile = useIsMobile();

  return isMobile ? (
    <MobileNav />
  ) : (
    <>
      <header className="nav__header">
        <div className="nav__item-logo">
          <Logo />
        </div>

        <nav className="nav">
          <Menubar>
            <LeftLayout layout={layout} />
            <UserMenu />
          </Menubar>
        </nav>
      </header>
    </>
  );
};

Nav.propTypes = {
  layout: PropTypes.oneOf(['dashboard', 'project'])
};

Nav.defaultProps = {
  layout: 'project'
};

const LeftLayout = (props) => {
  switch (props.layout) {
    case 'dashboard':
      return <DashboardMenu />;
    case 'project':
    default:
      return <ProjectMenu />;
  }
};

LeftLayout.propTypes = {
  layout: PropTypes.oneOf(['dashboard', 'project'])
};

LeftLayout.defaultProps = {
  layout: 'project'
};

const UserMenu = () => {
  const isLoginEnabled = getConfig('LOGIN_ENABLED');
  const isAuthenticated = useSelector(getAuthenticated);

  if (isLoginEnabled && isAuthenticated) {
    return <AuthenticatedUserMenu />;
  } else if (isLoginEnabled && !isAuthenticated) {
    return <UnauthenticatedUserMenu />;
  }

  return null;
};

const Logo = () => {
  const { t } = useTranslation();
  const user = useSelector((state) => state.user);

  if (user?.username) {
    return (
      <Link to={`/${user.username}/sketches`}>
        <LogoIcon
          role="img"
          aria-label={t('Common.p5logoARIA')}
          focusable="false"
          className="svg__logo"
        />
      </Link>
    );
  }

  return (
    <a href="https://p5js.org">
      <LogoIcon
        role="img"
        aria-label={t('Common.p5logoARIA')}
        focusable="true"
        className="svg__logo"
      />
    </a>
  );
};

const DashboardMenu = () => {
  const { t } = useTranslation();
  const editorLink = useSelector(selectSketchPath);
  return (
    <ul className="nav__items-left" role="group">
      <li className="nav__item nav__item--no-icon">
        <Link to={editorLink} className="nav__back-link">
          <CaretLeftIcon
            className="nav__back-icon"
            focusable="false"
            aria-hidden="true"
          />
          <span className="nav__item-header">{t('Nav.BackEditor')}</span>
        </Link>
      </li>
    </ul>
  );
};

const ProjectMenu = () => {
  const isUserOwner = useSelector(getIsUserOwner);
  const project = useSelector((state) => state.project);
  const user = useSelector((state) => state.user);

  const isUnsaved = !project?.id;

  const rootFile = useSelector(selectRootFile);

  const cmRef = useContext(CmControllerContext);

  const dispatch = useDispatch();

  const { t } = useTranslation();
  const {
    newSketch,
    saveSketch,
    downloadSketch,
    shareSketch
  } = useSketchActions();

  const replaceCommand =
    metaKey === 'Ctrl' ? `${metaKeyName}+H` : `${metaKeyName}+⌥+F`;
  const newFileCommand =
    metaKey === 'Ctrl' ? `${metaKeyName}+Alt+N` : `${metaKeyName}+⌥+N`;

  return (
    <ul className="nav__items-left" role="group">
      <MenubarSubmenu id="file" title={t('Nav.File.Title')}>
        <MenubarItem id="file-new" onClick={newSketch}>
          {t('Nav.File.New')}
        </MenubarItem>
        <MenubarItem
          id="file-save"
          isDisabled={
            !user.authenticated ||
            !getConfig('LOGIN_ENABLED') ||
            (project?.owner && !isUserOwner)
          }
          onClick={() => saveSketch(cmRef.current)}
        >
          {t('Common.Save')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+S</span>
        </MenubarItem>
        <MenubarItem
          id="file-duplicate"
          isDisabled={isUnsaved || !user.authenticated}
          onClick={() => dispatch(cloneProject())}
        >
          {t('Nav.File.Duplicate')}
        </MenubarItem>
        <MenubarItem
          id="file-share"
          isDisabled={isUnsaved}
          onClick={shareSketch}
        >
          {t('Nav.File.Share')}
        </MenubarItem>
        <MenubarItem
          id="file-download"
          isDisabled={isUnsaved}
          onClick={downloadSketch}
        >
          {t('Nav.File.Download')}
        </MenubarItem>
        <MenubarItem
          id="file-open"
          isDisabled={!user.authenticated}
          href={`/${user.username}/sketches`}
        >
          {t('Nav.File.Open')}
        </MenubarItem>
        <MenubarItem
          id="file-add-to-collection"
          isDisabled={
            !getConfig('UI_COLLECTIONS_ENABLED') ||
            !user.authenticated ||
            isUnsaved
          }
          href={`/${user.username}/sketches/${project?.id}/add-to-collection`}
        >
          {t('Nav.File.AddToCollection')}
        </MenubarItem>
        <MenubarItem
          id="file-examples"
          isDisabled={!getConfig('EXAMPLES_ENABLED')}
          href="/p5/sketches"
        >
          {t('Nav.File.Examples')}
        </MenubarItem>
      </MenubarSubmenu>
      <MenubarSubmenu id="edit" title={t('Nav.Edit.Title')}>
        <MenubarItem id="edit-tidy" onClick={cmRef.current?.tidyCode}>
          {t('Nav.Edit.TidyCode')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+Shift+F</span>
        </MenubarItem>
        <MenubarItem id="edit-find" onClick={cmRef.current?.showFind}>
          {t('Nav.Edit.Find')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+F</span>
        </MenubarItem>
        <MenubarItem id="edit-replace" onClick={cmRef.current?.showReplace}>
          {t('Nav.Edit.Replace')}
          <span className="nav__keyboard-shortcut">{replaceCommand}</span>
        </MenubarItem>
      </MenubarSubmenu>
      <MenubarSubmenu id="sketch" title={t('Nav.Sketch.Title')}>
        <MenubarItem
          id="sketch-add-file"
          onClick={() => dispatch(newFile(rootFile.id))}
        >
          {t('Nav.Sketch.AddFile')}
          <span className="nav__keyboard-shortcut">{newFileCommand}</span>
        </MenubarItem>
        <MenubarItem
          id="sketch-add-folder"
          onClick={() => dispatch(newFolder(rootFile.id))}
        >
          {t('Nav.Sketch.AddFolder')}
        </MenubarItem>
        <MenubarItem id="sketch-run" onClick={() => dispatch(startSketch())}>
          {t('Nav.Sketch.Run')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+Enter</span>
        </MenubarItem>
        <MenubarItem id="sketch-stop" onClick={() => dispatch(stopSketch())}>
          {t('Nav.Sketch.Stop')}
          <span className="nav__keyboard-shortcut">
            Shift+{metaKeyName}+Enter
          </span>
        </MenubarItem>
      </MenubarSubmenu>
      <MenubarSubmenu id="help" title={t('Nav.Help.Title')}>
        <MenubarItem
          id="help-shortcuts"
          onClick={() => dispatch(showKeyboardShortcutModal())}
        >
          {t('Nav.Help.KeyboardShortcuts')}
        </MenubarItem>
        <MenubarItem id="help-reference" href="https://p5js.org/reference/">
          {t('Nav.Help.Reference')}
        </MenubarItem>
        <MenubarItem id="help-about" href="/about">
          {t('Nav.Help.About')}
        </MenubarItem>
      </MenubarSubmenu>
      {getConfig('TRANSLATIONS_ENABLED') && <LanguageMenu />}
    </ul>
  );
};

const LanguageMenu = () => {
  const language = useSelector((state) => state.preferences.language);
  const dispatch = useDispatch();

  function handleLangSelection(event) {
    dispatch(setLanguage(event.target.value));
    dispatch(showToast('Toast.LangChange'));
  }

  return (
    <MenubarSubmenu
      id="lang"
      title={languageKeyToLabel(language)}
      triggerRole="button"
      listRole="listbox"
    >
      {sortBy(availableLanguages).map((key) => (
        // eslint-disable-next-line react/jsx-no-bind
        <MenubarItem
          id={key}
          key={key}
          value={key}
          onClick={handleLangSelection}
          role="option"
          selected={key === language}
        >
          {languageKeyToLabel(key)}
        </MenubarItem>
      ))}
    </MenubarSubmenu>
  );
};

const UnauthenticatedUserMenu = () => {
  const { t } = useTranslation();
  return (
    <ul className="nav__items-right" title="user-menu" role="group">
      <li className="nav__item">
        <Link to="/login" className="nav__auth-button">
          <span className="nav__item-header" title="Login">
            {t('Nav.Login')}
          </span>
        </Link>
      </li>
      <li className="nav__item-or" role="presentation">
        {t('Nav.LoginOr')}
      </li>
      <li className="nav__item">
        <Link to="/signup" className="nav__auth-button">
          <span className="nav__item-header" title="SignUp">
            {t('Nav.SignUp')}
          </span>
        </Link>
      </li>
    </ul>
  );
};

const AuthenticatedUserMenu = () => {
  const username = useSelector((state) => state.user.username);

  const { t } = useTranslation();
  const dispatch = useDispatch();

  return (
    <ul className="nav__items-right" title="user-menu" role="group">
      <MenubarSubmenu
        id="account"
        title={
          <span>
            {t('Nav.Auth.Hello')}, {username}!
          </span>
        }
      >
        <MenubarItem id="account-sketches" href={`/${username}/sketches`}>
          {t('Nav.Auth.MySketches')}
        </MenubarItem>
        <MenubarItem
          id="account-collections"
          href={`/${username}/collections`}
          isDisabled={!getConfig('UI_COLLECTIONS_ENABLED')}
        >
          {t('Nav.Auth.MyCollections')}
        </MenubarItem>
        <MenubarItem id="account-assets" href={`/${username}/assets`}>
          {t('Nav.Auth.MyAssets')}
        </MenubarItem>
        <MenubarItem id="account-settings" href="/account">
          {t('Preferences.Settings')}
        </MenubarItem>
        <MenubarItem id="account-logout" onClick={() => dispatch(logoutUser())}>
          {t('Nav.Auth.LogOut')}
        </MenubarItem>
      </MenubarSubmenu>
    </ul>
  );
};

export default Nav;
