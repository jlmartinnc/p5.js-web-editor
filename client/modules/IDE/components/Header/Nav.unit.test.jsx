/* eslint-disable react/prop-types */
import React from 'react';
import { reduxRender } from '../../../../test-utils';

import Nav from './Nav';

jest.mock('../../../../utils/generateRandomName');

// mock Menubar
jest.mock(
  '../../../../components/Menubar/Menubar',
  () =>
    function Menubar({ children, className = 'nav__menubar' }) {
      return (
        <ul className={className} role="menubar">
          {children}
        </ul>
      );
    }
);

// mock MenubarSubmenu
jest.mock('../../../../components/Menubar/MenubarSubmenu', () => {
  function MenubarSubmenu({ children, title }) {
    return (
      <li className="nav__item">
        <span role="menuitem">{title}</span>
        <ul role="menu" aria-label={`${title} menu`}>
          {children}
        </ul>
      </li>
    );
  }

  MenubarSubmenu.useMenuProps = () => ({
    isOpen: false,
    handlers: {}
  });

  return MenubarSubmenu;
});

// mock MenubarItem
jest.mock(
  '../../../../components/Menubar/MenubarItem',
  () =>
    function MenubarItem({ children, hideIf }) {
      if (hideIf) return null;

      return <li>{children}</li>;
    }
);

describe('Nav', () => {
  it('renders editor version for desktop', () => {
    const { asFragment } = reduxRender(<Nav />, { mobile: false });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders editor version for mobile', () => {
    const { asFragment } = reduxRender(<Nav />, { mobile: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders dashboard version for desktop', () => {
    const { asFragment } = reduxRender(<Nav layout="dashboard" />, {
      mobile: false
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders dashboard version for mobile', () => {
    const { asFragment } = reduxRender(<Nav layout="dashboard" />, {
      mobile: true
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
