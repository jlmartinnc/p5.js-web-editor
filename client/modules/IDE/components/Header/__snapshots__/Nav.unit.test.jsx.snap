// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders dashboard version for desktop 1`] = `
<DocumentFragment>
  <header
    class="nav__header"
  >
    <div
      class="nav__item-logo"
    >
      <a
        href="https://p5js.org"
      >
        <test-file-stub
          aria-label="p5.js Logo"
          classname="svg__logo"
          focusable="true"
          role="img"
        />
      </a>
    </div>
    <nav
      class="nav"
    >
      <ul
        class="nav__menubar"
        role="menubar"
      >
        <ul
          class="nav__items-left"
          role="group"
        >
          <li
            class="nav__item nav__item--no-icon"
          >
            <a
              class="nav__back-link"
              href="/"
            >
              <test-file-stub
                aria-hidden="true"
                classname="nav__back-icon"
                focusable="false"
              />
              <span
                class="nav__item-header"
              >
                Back to Editor
              </span>
            </a>
          </li>
        </ul>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Nav renders dashboard version for mobile 1`] = `
<DocumentFragment>
  .c4.c4.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #333;
  cursor: pointer;
  border: none;
  line-height: 1;
}

.c4.c4.c4 svg * {
  fill: #333;
}

.c4.c4.c4:disabled {
  cursor: not-allowed;
}

.c4.c4.c4 > * + * {
  margin-left: 0.6666666666666666rem;
}

.c5 {
  width: 4rem;
}

.c5 > svg {
  width: 100%;
  height: 100%;
}

.c0 {
  background: #F0F0F0;
  color: #333;
  padding: 0.6666666666666666rem 0;
  gap: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8333333333333334rem;
  box-shadow: #00000030 0px 2px 8px 0px;
  z-index: 10;
}

.c1 {
  width: 2.3333333333333335rem;
  aspect-ratio: 1;
  margin-left: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > svg {
  width: 100%;
  height: 100%;
}

.c1 > svg > path {
  fill: #ed225d;
  stroke: #ed225d;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
}

.c2 * {
  padding: 0;
  margin: 0;
}

.c2 > h5 {
  font-size: 1.0833333333333333rem;
  font-weight: normal;
}

.c3 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 ul.opened {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
  max-width: 90vw;
  max-height: 90vh;
  overflow-y: auto;
}

.c3 > div {
  position: relative;
}

.c3 > div ul {
  foreground: #000;
  background: #F0F0F0;
  border: #A6A6A6;
  position: absolute;
  overflow: hidden;
  z-index: 10;
  right: 10px;
  -webkit-transform: translateX(0.5rem);
  -ms-transform: translateX(0.5rem);
  transform: translateX(0.5rem);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  opacity: 0;
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: opacity 100ms,-webkit-transform 100ms 200ms;
  -webkit-transition: opacity 100ms,transform 100ms 200ms;
  transition: opacity 100ms,transform 100ms 200ms;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
  box-shadow: rgba(60,64,67,0.3) 0px 1px 2px 0px,rgba(60,64,67,0.15) 0px 1px 3px 1px;
  min-width: 10rem;
  border-radius: 0.16666666666666666rem;
}

.c3 > div ul li.user {
  color: #2D67F6;
  font-size: 1.3333333333333333rem;
  font-weight: bold;
  margin: 0.5rem 1rem;
}

.c3 > div ul > b {
  margin: 0.5rem 1rem;
  position: relative;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
}

.c3 > div ul > b::after {
  content: '';
  position: absolute;
  opacity: calc(0.5);
  top: 50%;
  -webkit-transform: translate(100%,-50%);
  -ms-transform: translate(100%,-50%);
  transform: translate(100%,-50%);
  right: -0.5rem;
  width: 4.583333333333333rem;
  height: 1px;
  background-color: #4D4D4D;
}

.c3 > div ul > li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 > div ul > li a,
.c3 > div ul > li button {
  width: 100%;
  padding: 0.6666666666666666rem 1.25rem 0.6666666666666666rem 0.8333333333333334rem;
  font-size: 1.5rem;
  text-align: left;
}

.c3 > div ul > li a:hover,
.c3 > div ul > li button:hover {
  background-color: #ed225d;
  color: #FFF;
}

<ul
    class="c0"
    role="menubar"
  >
    <div
      class="c1"
    >
      <test-file-stub />
    </div>
    <div
      class="c2"
    >
      <h1>
        <span
          class="editable-input editable-input--is-not-editing editable-input--has-value "
        >
          <button
            aria-hidden="false"
            aria-label="Edit sketch name"
            class="editable-input__label"
          >
            <span>
              Test project name
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="editable-input__icon"
              focusable="false"
            />
          </button>
          <input
            aria-hidden="true"
            aria-label="New sketch name"
            class="editable-input__input"
            disabled=""
            maxlength="128"
            type="text"
            value="Test project name"
          />
        </span>
      </h1>
    </div>
    <div
      class="c3"
    >
      <div>
        <a
          href="/login"
        >
          <button
            class="c4 c5"
            display="inline"
            focusable="false"
            kind="primary"
            type="button"
          >
            <test-file-stub
              aria-hidden="true"
              classname="icons__StyledIcon-sc-xmer15-0 dStXqm"
              focusable="false"
            />
          </button>
        </a>
      </div>
      <div>
        <button
          class="c4 c5"
          display="inline"
          focusable="false"
          kind="primary"
          type="button"
        >
          <test-file-stub
            aria-hidden="true"
            classname="icons__StyledIcon-sc-xmer15-0 dStXqm"
            focusable="false"
          />
        </button>
        <ul
          class=""
        >
          <b>
            File
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              New
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Save
            </button>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="/p5/sketches"
            >
              Examples
            </a>
          </li>
          <b>
            Edit
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Tidy Code
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Find
            </button>
          </li>
          <b>
            Sketch
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Add File
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Add Folder
            </button>
          </li>
          <b>
            Settings
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Preferences
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Language
            </button>
          </li>
          <b>
            Help
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Keyboard Shortcuts
            </button>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="https://p5js.org/reference/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Reference
            </a>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="/about"
            >
              About
            </a>
          </li>
        </ul>
      </div>
    </div>
  </ul>
</DocumentFragment>
`;

exports[`Nav renders editor version for desktop 1`] = `
<DocumentFragment>
  <header
    class="nav__header"
  >
    <div
      class="nav__item-logo"
    >
      <a
        href="https://p5js.org"
      >
        <test-file-stub
          aria-label="p5.js Logo"
          classname="svg__logo"
          focusable="true"
          role="img"
        />
      </a>
    </div>
    <nav
      class="nav"
    >
      <ul
        class="nav__menubar"
        role="menubar"
      >
        <ul
          class="nav__items-left"
          role="group"
        >
          <li
            class="nav__item"
          >
            <span
              role="menuitem"
            >
              File
            </span>
            <ul
              aria-label="File menu"
              role="menu"
            >
              <li>
                New
              </li>
              <li>
                Save
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+S
                </span>
              </li>
              <li>
                Duplicate
              </li>
              <li>
                Share
              </li>
              <li>
                Download
              </li>
              <li>
                Open
              </li>
              <li>
                Add to Collection
              </li>
              <li>
                Examples
              </li>
            </ul>
          </li>
          <li
            class="nav__item"
          >
            <span
              role="menuitem"
            >
              Edit
            </span>
            <ul
              aria-label="Edit menu"
              role="menu"
            >
              <li>
                Tidy Code
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+Shift+F
                </span>
              </li>
              <li>
                Find
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+F
                </span>
              </li>
              <li>
                Replace
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+H
                </span>
              </li>
            </ul>
          </li>
          <li
            class="nav__item"
          >
            <span
              role="menuitem"
            >
              Sketch
            </span>
            <ul
              aria-label="Sketch menu"
              role="menu"
            >
              <li>
                Add File
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+Alt+N
                </span>
              </li>
              <li>
                Add Folder
              </li>
              <li>
                Run
                <span
                  class="nav__keyboard-shortcut"
                >
                  Ctrl+Enter
                </span>
              </li>
              <li>
                Stop
                <span
                  class="nav__keyboard-shortcut"
                >
                  Shift+Ctrl+Enter
                </span>
              </li>
            </ul>
          </li>
          <li
            class="nav__item"
          >
            <span
              role="menuitem"
            >
              Help
            </span>
            <ul
              aria-label="Help menu"
              role="menu"
            >
              <li>
                Keyboard Shortcuts
              </li>
              <li>
                Reference
              </li>
              <li>
                About
              </li>
            </ul>
          </li>
        </ul>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Nav renders editor version for mobile 1`] = `
<DocumentFragment>
  .c4.c4.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #333;
  cursor: pointer;
  border: none;
  line-height: 1;
}

.c4.c4.c4 svg * {
  fill: #333;
}

.c4.c4.c4:disabled {
  cursor: not-allowed;
}

.c4.c4.c4 > * + * {
  margin-left: 0.6666666666666666rem;
}

.c5 {
  width: 4rem;
}

.c5 > svg {
  width: 100%;
  height: 100%;
}

.c0 {
  background: #F0F0F0;
  color: #333;
  padding: 0.6666666666666666rem 0;
  gap: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8333333333333334rem;
  box-shadow: #00000030 0px 2px 8px 0px;
  z-index: 10;
}

.c1 {
  width: 2.3333333333333335rem;
  aspect-ratio: 1;
  margin-left: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > svg {
  width: 100%;
  height: 100%;
}

.c1 > svg > path {
  fill: #ed225d;
  stroke: #ed225d;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
}

.c2 * {
  padding: 0;
  margin: 0;
}

.c2 > h5 {
  font-size: 1.0833333333333333rem;
  font-weight: normal;
}

.c3 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 ul.opened {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
  max-width: 90vw;
  max-height: 90vh;
  overflow-y: auto;
}

.c3 > div {
  position: relative;
}

.c3 > div ul {
  foreground: #000;
  background: #F0F0F0;
  border: #A6A6A6;
  position: absolute;
  overflow: hidden;
  z-index: 10;
  right: 10px;
  -webkit-transform: translateX(0.5rem);
  -ms-transform: translateX(0.5rem);
  transform: translateX(0.5rem);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  opacity: 0;
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: opacity 100ms,-webkit-transform 100ms 200ms;
  -webkit-transition: opacity 100ms,transform 100ms 200ms;
  transition: opacity 100ms,transform 100ms 200ms;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
  box-shadow: rgba(60,64,67,0.3) 0px 1px 2px 0px,rgba(60,64,67,0.15) 0px 1px 3px 1px;
  min-width: 10rem;
  border-radius: 0.16666666666666666rem;
}

.c3 > div ul li.user {
  color: #2D67F6;
  font-size: 1.3333333333333333rem;
  font-weight: bold;
  margin: 0.5rem 1rem;
}

.c3 > div ul > b {
  margin: 0.5rem 1rem;
  position: relative;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
}

.c3 > div ul > b::after {
  content: '';
  position: absolute;
  opacity: calc(0.5);
  top: 50%;
  -webkit-transform: translate(100%,-50%);
  -ms-transform: translate(100%,-50%);
  transform: translate(100%,-50%);
  right: -0.5rem;
  width: 4.583333333333333rem;
  height: 1px;
  background-color: #4D4D4D;
}

.c3 > div ul > li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 > div ul > li a,
.c3 > div ul > li button {
  width: 100%;
  padding: 0.6666666666666666rem 1.25rem 0.6666666666666666rem 0.8333333333333334rem;
  font-size: 1.5rem;
  text-align: left;
}

.c3 > div ul > li a:hover,
.c3 > div ul > li button:hover {
  background-color: #ed225d;
  color: #FFF;
}

<ul
    class="c0"
    role="menubar"
  >
    <div
      class="c1"
    >
      <test-file-stub />
    </div>
    <div
      class="c2"
    >
      <h1>
        <span
          class="editable-input editable-input--is-not-editing editable-input--has-value "
        >
          <button
            aria-hidden="false"
            aria-label="Edit sketch name"
            class="editable-input__label"
          >
            <span>
              Test project name
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="editable-input__icon"
              focusable="false"
            />
          </button>
          <input
            aria-hidden="true"
            aria-label="New sketch name"
            class="editable-input__input"
            disabled=""
            maxlength="128"
            type="text"
            value="Test project name"
          />
        </span>
      </h1>
    </div>
    <div
      class="c3"
    >
      <div>
        <a
          href="/login"
        >
          <button
            class="c4 c5"
            display="inline"
            focusable="false"
            kind="primary"
            type="button"
          >
            <test-file-stub
              aria-hidden="true"
              classname="icons__StyledIcon-sc-xmer15-0 dStXqm"
              focusable="false"
            />
          </button>
        </a>
      </div>
      <div>
        <button
          class="c4 c5"
          display="inline"
          focusable="false"
          kind="primary"
          type="button"
        >
          <test-file-stub
            aria-hidden="true"
            classname="icons__StyledIcon-sc-xmer15-0 dStXqm"
            focusable="false"
          />
        </button>
        <ul
          class=""
        >
          <b>
            File
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              New
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Save
            </button>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="/p5/sketches"
            >
              Examples
            </a>
          </li>
          <b>
            Edit
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Tidy Code
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Find
            </button>
          </li>
          <b>
            Sketch
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Add File
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Add Folder
            </button>
          </li>
          <b>
            Settings
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Preferences
            </button>
          </li>
          <li>
            <button
              aria-disabled="false"
            >
              Language
            </button>
          </li>
          <b>
            Help
          </b>
          <li>
            <button
              aria-disabled="false"
            >
              Keyboard Shortcuts
            </button>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="https://p5js.org/reference/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Reference
            </a>
          </li>
          <li>
            <a
              aria-disabled="false"
              href="/about"
            >
              About
            </a>
          </li>
        </ul>
      </div>
    </div>
  </ul>
</DocumentFragment>
`;
